### ISSUE TITLE
binary classification: Sequential.evaluate() get different accuracy than Sequential.predict()

### ISSUE THREAD

**Issue by llcao** [fchollet/keras#2108]
I modified the example (MNIST-MLP)  to a binary classification model, and got 1.0 validation accuracy using model.evaluate().   

However, if I use model.predict() and calculate the accuracy by myself, the accuracy is 0.997. I think this is more than rounding error but could not figure out how to fix.

I am attaching my code below. 

```
from __future__ import print_function
import numpy as np
np.random.seed(1337)  # for reproducibility

from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD, Adam, RMSprop
from keras.utils import np_utils


batch_size = 128
nb_classes = 10
nb_epoch = 2

# the data, shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

X_train = X_train.reshape(60000, 784)
X_test = X_test.reshape(10000, 784)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')


binary=True
if not binary: # convert class vectors to binary class matrices
    Y_train = np_utils.to_categorical(y_train, nb_classes)
    Y_test = np_utils.to_categorical(y_test, nb_classes)
else: # my modification as binary classification
    c = 1
    Y_train = (y_train==c).astype(np.int32)
    Y_test = (y_test==c).astype(np.int32)

model = Sequential()
model.add(Dense(512, input_shape=(784,)))
model.add(Activation('relu'))
model.add(Dropout(0.2))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.2))

if not binary: 
    model.add(Dense(10))
    model.add(Activation('softmax'))
    rms = RMSprop()
    model.compile(loss='categorical_crossentropy', optimizer=rms)

else:
    model.add(Dense(1))
    model.add(Activation('sigmoid'))
    rms = RMSprop()
    model.compile(loss='binary_crossentropy', optimizer=rms)


model.fit(X_train, Y_train,
          batch_size=batch_size, nb_epoch=nb_epoch,
          show_accuracy=True, verbose=2,
          validation_data=(X_test, Y_test))
score = model.evaluate(X_test, Y_test,
                       show_accuracy=True, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

if binary:
    esti = model.predict(X_test)
    estiy = (esti>0.5)
    acc = np.mean(estiy[:,0] == Y_test)
    print ('my acc:', acc)

```

**Comment by tboquet** [comment]
You should try with 10 outputs for the sigmoid if you want unnormalized probabilities:
```python
...
else:
    model.add(Dense(10))
    model.add(Activation('sigmoid'))
    rms = RMSprop()
    model.compile(loss='binary_crossentropy', optimizer=rms)
```

In this example, it does not make a lot of sense since classes are mutually exclusive but you could apply it on another problem.

**Comment by llcao** [comment]
Hmm, note that I already change the data into binary setting. So I think Dense(1) is reasonable. 
See the code:
```python
else: # my modification as binary classification
    c = 1
    Y_train = (y_train==c).astype(np.int32)
    Y_test = (y_test==c).astype(np.int32)
```

**Comment by tboquet** [comment]
Really sorry! I thought it was a conversion between one-hot encoding to binary bug since it has the same behavior.

With Theano on GPU I have these results:

```python
Train on 60000 samples, validate on 10000 samples
Epoch 1/2
1s - loss: 0.0288 - acc: 0.9908 - val_loss: 0.0114 - val_acc: 0.9967
Epoch 2/2
1s - loss: 0.0130 - acc: 0.9961 - val_loss: 0.0086 - val_acc: 0.9974
Test score: 0.00856090349201
Test accuracy: 0.9974
my acc: 0.9974
```

This is my theano config:

```
[global]
device=gpu
floatX=float32
[nvcc]\nfastmath=True
```

Are you running your code on GPU/CPU Theano/Tensorflow?
